.github/workflows/reusable-workflow.yaml

```diff
@@ -0,0 +1,127 @@
+        run: |
+          npm install -g typescript
+      - uses: actions/setup-java@v4
+        with:
+          distribution: 'zulu'
+          java-version: '21'
+      - name: Setup Android SDK
+        uses: android-actions/setup-android@v3
+      - uses: gradle/actions/setup-gradle@v4
+        with:
+          gradle-version: 8.7
+      - name: Run Lint
+        if: ${{ github.event_name }} == 'pull_request'
+        run: |
+          ./gradlew libs:${{ inputs.lib }}:lint
+      - uses: ruby/setup-ruby@v1
+        with:
+          ruby-version: '3.2'
+          bundler-cache: true
+      - name: Report Static Analysis
+        if: ${{ github.event_name }} == 'pull_request'
+        env:
+          DANGER_GITHUB_API_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          LIB: ${{ inputs.lib }}
+        run: bundle exec danger --dangerfile=.github/DangerFiles/StaticAnalysis.rb --danger_id="${{ inputs.lib }}"
+      - name: Build for Testing
+        if: success() || failure()
+        run: |
+          ./gradlew libs:${{ inputs.lib }}:assembleAndroidTest
+          ./gradlew native:NativeSampleApps:RestExplorer:assembleDebug
+      - uses: 'google-github-actions/auth@v2'
+        if: success() || failure()
+        with:
+          credentials_json: '${{ secrets.GCLOUD_SERVICE_KEY }}'
+      - uses: 'google-github-actions/setup-gcloud@v2'
+        if: success() || failure()
+      - name: Run Tests
+        if: success() || failure()
+        run: |
+          # Most used according to https://gs.statcounter.com/android-version-market-share/mobile-tablet/worldwide
+          DEFAULT_API_VERSION="34" 
+          FULL_API_RANGE="28 29 30 31 32 33 34 35"
+
+          DEVICES=""
+          if [[ ${{ github.event_name == 'pull_request' }} ]]; then
+            DEVICES="--device model=MediumPhone.arm,version=${DEFAULT_API_VERSION},locale=en,orientation=portrait "
+          else
+            for LEVEL in $FULL_API_RANGE
+            do
+                DEVICES+="--device model=MediumPhone.arm,version=${LEVEL},locale=en,orientation=portrait "
+            done
+          fi
+
+          COMMAND="gcloud firebase test android run "
+          COMMAND+="--project mobile-apps-firebase-test "
+          COMMAND+="--type instrumentation "
+          COMMAND+="--app \"native/NativeSampleApps/RestExplorer/build/outputs/apk/debug/RestExplorer-debug.apk\" "
+          COMMAND+="--test=libs/${{ inputs.lib }}/build/outputs/apk/androidTest/debug/${{ inputs.lib }}-debug-androidTest.apk "
+          COMMAND+="${DEVICES}"
+          COMMAND+="--environment-variables coverage=true,coverageFile=\"/sdcard/coverage.ec\" "
+          COMMAND+="--directories-to-pull=/sdcard  "
+          COMMAND+="--results-dir=${{ inputs.lib }}-${{github.run_number}} "
+          COMMAND+="--results-history-name=${{ inputs.lib }} "
+          COMMAND+="--timeout=20m --no-auto-google-login --no-record-video --no-performance-metrics --num-flaky-test-attempts=1"
+
+          eval "$COMMAND"
```

Comment:
I don't love this implementation, but with the above section to add a single device for PRs and a device per API level for nightly runs I am not sure what else I can do.  It's still pretty readable so I'd rather not break it out into yet another ruby or JS script.